dsc/workflows/opencourseware.py

```diff
@@ -0,0 +1,203 @@
+import json
+import logging
+import zipfile
+from collections import defaultdict
+from collections.abc import Iterator
+from typing import Any
+
+import smart_open
+
+from dsc.utilities.aws.s3 import S3Client
+from dsc.workflows.base import Workflow
+
+logger = logging.getLogger(__name__)
+
+
+class OpenCourseWare(Workflow):
+    """Workflow for OpenCourseWare (OCW) deposits.
+
+    The deposits managed by this workflow are requested by
+    Technical Services staff and are for submission to DSpace@MIT.
+    """
+
+    workflow_name: str = "opencourseware"
+    metadata_mapping_path: str = "dsc/workflows/metadata_mapping/opencourseware.json"
+
+    def reconcile_bitstreams_and_metadata(self) -> tuple[set[str], set[str]]:
+        """Reconcile bitstreams against metadata.
+
+        Generate a list of bitstreams without item identifiers and item identifiers
+        without bitstreams. Any discrepancies will be addressed by the engineer and
+        stakeholders as necessary.
+
+        For OpenCourseWare deposits, the zip files are the bitstreams to be deposited
+        into DSpace, but they also must contain a 'data.json' file, which is the metadata.
+        As such, the 'reconcile' method only determines whether there are any
+        bitstreams without metadata (any zip files without a 'data.json').
+        Metadata without bitstreams are basically impossible because the metadata
+        ('data.json') is inside the bitstream (zip file), hence an empty set is returned.
+        """
+        bitstream_dict = self._build_bitstream_dict()
+
+        # extract item identifiers from bitstream_dict
+        item_identifiers = list(bitstream_dict.keys())
+
+        bitstreams_without_metadata = set(item_identifiers) - set(
+            self._identify_bitstreams_with_metadata(item_identifiers)
+        )
+        item_metadata_without_bitstreams: set[str] = set()
+        return item_metadata_without_bitstreams, bitstreams_without_metadata
+
+    def _build_bitstream_dict(self) -> dict:
+        """Build dictionary of bitstreams.
+
+        This method will look for zip files within the designated 'batch' folder
+        of the S3 bucket (i.e., self.batch_path). In the case of OpenCourseWare
+        deposits, this method expects:
+            * a single (1) zipped file per item identifier
+            * filename of zipped file must correspond to item identifier
+              (i.e., '<item_identifier>'.zip)
+        """
+        s3_client = S3Client()
+        bitstreams = list(
+            s3_client.files_iter(
+                bucket=self.s3_bucket, prefix=self.batch_path, file_type=".zip"
+            )
+        )
+        bitstream_dict: dict[str, list[str]] = defaultdict(list)
+        for bitstream in bitstreams:
+            file_name = bitstream.split("/")[-1]
+            item_identifier = file_name.replace(".zip", "")
```

Comment:
Relatively minor and unlikely to crop up, but I'd be cautious about using `.replace()` in situations like this.  If the filename was something like `mr.zips-amazing-class.zip` you'd get `mrs-amazing-class` as the identifier because both instances of `.zip` were removed:

```python
x = "mr.zips-amazing-class.zip"
x.replace(".zip", "")
Out[2]: 'mrs-amazing-class'
```

I'd suggest something like `.split(".")[0]` but that is also problematic if there is a period `.` anywhere in the filename.

SO, taken altogether, feels like `.removeprefix(".zip")` could be the best bet here.

Reply:
Ah, thank you for this suggestion! I will use these methods (`removesuffix` in this case) moving forward.